### ISSUE TITLE
JIT: DifferentiableGraph/Requires grad handled badly by ProfilingExecutor and fuser fallbacks

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#49299]
## 🐛 Bug

To me, it seems like the interplay between requiring gradients and the profiling executor/fuser fallback seems not ideal:
1) in the fuser1 (TensorExpr) and fuser2 (CudaFuser), if I run a few grad-requiring tensors through my model and then run non-grad-requiring, it will run into the DifferentiableSubgraph and give me a grad-requiring output even though none of the inputs requires gradient. (It will also run a kernel outputting intermediates that we don't need, so perf loss...)
2) in the fuser2 (CudaFuser) the CudaFusionGuard doesn't check for requires grad, so if I run a few non-grad-requiring tensors first, and feed a grad-requiring input, it'll drop the gradient (this is different for fuser1 where the typecheck checks for requires grad and then drops into the fallback).

To me it would seem that if we want to pull off the fallback stuff, we would properly guard DifferentiableGraph with what requires grad and the fusion guards with nothing requires grad (we might also have cases where which gradients are required varies).

## To Reproduce

```python
for fuser in ["fuser1", "fuser2"]:
    for rq in [True, False]:
        c = torch.jit.fuser(fuser)
        c.__enter__()

        def ratio_iou(x1, y1, w1, h1, x2, y2, w2, h2):
            xi = torch.max(x1, x2)                                  # Intersection left
            yi = torch.max(y1, y2)                                  # Intersection top
            wi = torch.clamp(torch.min(x1+w1, x2+w2) - xi, min=0.)  # Intersection width
            hi = torch.clamp(torch.min(y1+h1, y2+h2) - yi, min=0.)  # Intersection height
            area_i = wi * hi                                        # Area Intersection
            area_u = w1 * h1 + w2 * h2 - wi * hi                    # Area Union
            return area_i / torch.clamp(area_u, min=1e-5)           # Intersection over Union

        ratio_iou_scripted = torch.jit.script(ratio_iou)

        x1, y1, w1, h1, x2, y2, w2, h2 = torch.randn(8, 100, 1000, device='cuda', requires_grad=not rq).exp()

        for i in range(10):
            ratio_iou_scripted.graph_for(x1, y1, w1, h1, x2, y2, w2, h2)
        #print(ratio_iou_scripted.graph_for(x1, y1, w1, h1, x2, y2, w2, h2))

        x1, y1, w1, h1, x2, y2, w2, h2 = torch.randn(8, 100, 1000, device='cuda', requires_grad=rq).exp()
        print(fuser, x1.requires_grad, ratio_iou_scripted(x1, y1, w1, h1, x2, y2, w2, h2).requires_grad)
```

## Expected behavior

`requires_grad` of output same as for input.

## Environment

PyTorch master.

## Additional context

This used to be handled by the ArgSpec mechanism.

cc @gmagogsfm

**Comment by t-vi** [comment]
@csarofeen @jjsjann123  @Krovatkin @eellison @ZolotukhinM

**Comment by t-vi** [comment]
I'd be happy to implement a Guard+Fallback for DifferentiableGraph and fix the nvFuser if we can agree that that's what we want.

**Comment by Krovatkin** [comment]
1. in the fuser1 (TensorExpr) and fuser2 (CudaFuser), if I run a few grad-requiring tensors through my model and then run non-grad-requiring, it will run into the DifferentiableSubgraph and give me a grad-requiring output even though none of the inputs requires gradient. (It will also run a kernel outputting intermediates that we don't need, so perf loss...)

@t-vi are you aware of some models where the perf loss due to intermediate tensor computations could be significant? I'm playing with a prototype that fixes this exact issue and I'd love to measure it against some real models. 
Note, I'm still running such graphs through DifferentiableGraphs because it seems that the overhead of nesting inside DifferentiableGraphs seems to be 2-3% on average even for very small tensors vs a possibly non-trivial amount of work.

**Comment by t-vi** [comment]
It's more a correctness thing to me and I didn't get to the real world models in my exploration, but so the IOU above does produce an awful lot of intermediates (this has been one of my standard examples since back when I discussed JIT and fusion for detection models with @fmassa in 2018).
But really, how to fix it would seem to also tie into @jjsjann123 and @eelison 's conversation on slack this week about checking which tensors require gradients in layernorm. This is what has me thinking that guarding the `DifferentiableGraph` on the exact pattern and eliminating unnecessary computation would be the way forward. @eelison mentioned that he had a prototype for this (at least that's how I understood it).

**Comment by jjsjann123** [comment]
> I'd be happy to implement a Guard+Fallback for DifferentiableGraph and fix the nvFuser if we can agree that that's what we want.

Guard+Fallback sounds very tempting. Slightly expanding this problem: we can guard more than just `requires_grad_`, maybe as well as branches from conditional statements. This could go beyond for just `DifferentiableGraph`, forward graph alone could also use this guard and simplify its control flow to expose more fusion opportunities.

**Comment by eellison** [comment]
Hi @t-vi, thanks for the issue write up! 

Agreed that since it's a correctness issue we should fix this regardless of it has a negligible perf improvement. It sounds like we're all pretty aligned as to a solution - guard DifferentiableGraph with what inputs require grad. 

A few questions for implementation:

Currently when we create the differentiable graph, the Diff Graph input values are not yet annotated with whether or not they are profiled with grad, because the profiles still exist. 

```
graph(%x.1 : Tensor):
  %1 : int = prim::Constant[value=1]()
  %4 : Tensor = prim::profile[profiled_type=Float(2, strides=[1], requires_grad=0, device=cpu)](%x.1)
  %5 : Tensor = prim::profile[profiled_type=Float(2, strides=[1], requires_grad=0, device=cpu)](%x.1)
  %2 : Tensor = aten::add(%4, %5, %1) # test/elias.py:5:11
  %6 : Tensor = prim::profile[profiled_type=Float(2, strides=[1], requires_grad=0, device=cpu)](%2)
  %7 : Tensor = prim::profile[profiled_type=Float(2, strides=[1], requires_grad=0, device=cpu)](%x.1)
  %3 : Tensor = aten::add(%6, %7, %1) # test/elias.py:5:11
  %8 : Tensor = prim::profile[profiled_type=Float(2, strides=[1], requires_grad=0, device=cpu)](%3)
   = prim::profile()
  return (%8)
``` 
There are a few solutions here: 
- Erase the profiled nodes and set the type. I don't particularly like this, because the profiled types are not guaranteed to be seen in execution. They are more suggestions. That means any pass that runs in the optimization pipeline can't optimize on tensor type, and if it does it's a bug. I had a potential solution here that I profiled with that would set `profiled` types 

 `%2 : Profiled[Float(2, strides=[1], requires_grad=0, device=cpu)] = aten::add(%4, %5, %1)`. Here, we're distinguishing between types which are suggestions, and which are guaranteed. When a pass like peephole accessed a `size()` type on the tensor type, if it were a profiled type it would return None. Within `tensorexpr_fuser`, we would set a guard to allow access to profiled types because the pass sets its own guards. This honestly ended up being kind of confusing and a bunch of work without that much gain, so I didnt try to land it.

- `CreateDifferentiableGraphs` can just look at the profiled uses and do its own aggregation of which inputs require grad, without having to remove `prim::profile` nodes. This seems easy to implement and straightforward to me. 

A related question is what should the differentiable graph guard on. Should it be just which inputs require grad, or the complete profiled tensor type? 

Part of the `Profiled[Float(2, ...)]` idea was to distinguish that the values within Differentiable Graphs had already been guarded on, so they would not require a redundant `if TypeCheck` node, and that we could make optimizations based on their types (`x.dim(), etc). I'm not in favor of this. 

The difficulty here is type of the inputs of the Differentiable Graph have been guaranteed because we have checked them, but any non-input value's type is not guaranteed unless we prove it. I'm also not convinced that it really speeds up perf or is needed except in a couple cases, where we already have concrete fixes in mind (layer_norm, relu). It also doesnt really compose well with https://github.com/pytorch/pytorch/pull/47434. So my vote is we don't do this.

> This could go beyond for just DifferentiableGraph, forward graph alone could also use this guard and simplify its control flow to expose more fusion opportunities

The difficulty with the `if (guard) { optimized } { unoptimized}` is that the optimizations do not really compose. Honestly we had a way of doing this so optimizations would compose and it was Bailouts. People said that cleaning up the graph didn't matter, and we scrapped it. If graph clean up is a huge issue IMO we should just do bailouts again but with more people working on it.

**Comment by t-vi** [comment]
I must admit I never got into the details of Bailouts vs. Guards and I don't have an opinion, I've been looking at Guards because the more knowledgeable people seem to use them...

> A related question is what should the differentiable graph guard on. Should it be just which inputs require grad, or the complete profiled tensor type?

My take on the Differentiable Graphs is that `requires_grad?` as a tensor type attribute has a lot less potential for surprises than anything shape.
- This would suggest that a mechanism guarding DifferentiableGraphs just on requires_grad would be fine.
- There may be corner cases where whether or not we have requires_grad changes shape (if from defined to undefined is likely quite legitimate, but people will probably do all sorts of funny things), so correctness wants us not to assume things about the nodes moved into the subgraph.
- The fusion guards could then just require requires_gradient to be false.

You probably are quite aware of it, but just in case: to me fusers are not the only reason to have DifferentiableGraphs but I'm seeing it a a more general mechanism in it (e.g. for mul / matmul where only one operand needs grad, but potentially other things with `grad_input_mask` in autograd which seem to be easily achieved for DCE). I'm wondering if things are stable enough if we suddenly got the idea to radically expand what can be put into a DifferentiableGraph.

**Comment by jjsjann123** [comment]
Thanks to @eellison for iterating options in great detail.

I'm quite on board with your idea. I think a generalized approach (guaranteed type vs suggestion type), as appealing as it may sound, does not necessarily put much on the table at this moment. The tricky part is to generalize them to be useful in different cases. If the only thing that we are dealing with at this moment are `requires_grad_`, it's hard to justify the added complexity.

Similarly, `TypeCheck` is a little too rigid just as you mentioned about the binding of `if (guard)` to `requires_grad_` to other type information as rank e.t.c..

A comment on the proposed `BailOut` approach. I'm quite a big fan of `BailOut` because it handles the cleanup. Versus customized guards where optimization pass writer has to manually hook up the fallback, having said that, I do think it's necessary to have the flexibility provided by customized guards.

I think our previous frustration with `BailOut` comes from its abundance in the graph (which blocks fusion opportunity across `BailOuts`), as well as its rigid check on static shapes e.t.c. (which does not suit the need for optimization targeting dynamic shapes).
In this specific problem, if we relax `BailOut` to check only the needed information (`requires_grad_`) and use it with caution (for now, only apply it to inputs to `DifferentiableGraph`). I don't see any major blocks with this hybrid approach.

**Comment by jjsjann123** [comment]
FYI, I've ran into autodiff issue before, where vjps produced in backward graph are not linked/attached to the right input tensor. https://github.com/pytorch/pytorch/pull/49430

Not sure if we are hitting the same thing tho, just link it here for a reference.

**Comment by t-vi** [comment]
I think they're distinct, but it is good to have that fixed, too.
